<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<title>My Project: Documentation for the PyRosetta Toolkit GUI</title>

<link href="tabs.css" rel="stylesheet" type="text/css"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />

<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>

</head>
<body>
<div id="top"><!-- do not remove this div! -->


<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  
  
  <td style="padding-left: 0.5em;">
   <div id="projectname">My Project
   
   </div>
   
  </td>
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.6.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
</div>
<div class="header">
  <div class="headertitle">
<div class="title">Documentation for the PyRosetta Toolkit GUI </div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"><dl class="author"><dt><b>Author:</b></dt><dd>Jared Adolf-Bryfogle</dd></dl>
<h2><a class="anchor" id="metadata"></a>
Metadata</h2>
<p>Last edited April 2013. Code by Jared Adolf-Bryfogle &lt;<a href="mailto:jadolfbr@gmail.com">jadolfbr@gmail.com</a>&gt;. The PI for this application is Roland Dunbrack &lt;<a href="mailto:Roland.Dunbrack@fccc.edu">Roland.Dunbrack@fccc.edu</a>&gt;</p>
<h2><a class="anchor" id="code_demos"></a>
Code and Demo</h2>
<p>The application is available in both rosetta_source/GUIs/pyrosetta_toolkit and in the newest PyRosetta binary releases.</p>
<p>PyRosetta Setup: </p>
<ul>
<li><a href="http://www.pyrosetta.org/faq#TOC-1.-How-do-I-obtain-and-install-PyRosetta-">http://www.pyrosetta.org/faq#TOC-1.-How-do-I-obtain-and-install-PyRosetta-</a></li>
</ul>
<p>Optional PyMOL Integration Setup: </p>
<ul>
<li>See <a href="http://www.pyrosetta.org/pymol_mover-tutorial">http://www.pyrosetta.org/pymol_mover-tutorial</a></li>
</ul>
<p>Optional SCWRL Integration Setup: </p>
<ul>
<li>Obtain SCWRL from: <a href="http://dunbrack.fccc.edu/scwrl4/">http://dunbrack.fccc.edu/scwrl4/</a> </li>
<li>Copy the exe/binary into the /GUIs/pyrosetta_toolkit/SCWRL directory in the respective OS directory.</li>
</ul>
<p>Run the program: </p>
<ul>
<li>Use <code>./pyrosetta_toolkit.py</code> within /GUIs/pyrosetta_toolkit in rosetta_source or the PyRosetta directory to run the application after setting up PyRosetta paths.</li>
</ul>
<h2><a class="anchor" id="refs"></a>
References</h2>
<p>Jared Adolf-Bryfogle and Roland Dunbrack (2013) "The PyRosetta Toolkit: A Graphical User Interface for the Rosetta Software Suite. " PloS-ONE RosettaCON2012 special collection (Submitted)</p>
<h2><a class="anchor" id="purpose"></a>
Purpose</h2>
<p>This GUI is used to setup Rosetta specific filetypes, do interactive modeling using Rosetta as its base, run rosetta protocols, and analyze decoy results.</p>
<h2><a class="anchor" id="limits"></a>
Limitations</h2>
<p>This application does not run all Rosetta Applications. It does, however, run the most common ones. Although multiprocessing is implemented, full production runs of many rosetta applications usually require a cluster. Always check relevant Rosetta Documentation on RosettaCommons. Options for all protocols can be added through the Option System Manager or through it's protocol-specific UI. Symmetry is not supported at this time.</p>
<h2><a class="anchor" id="menus"></a>
Menus</h2>
<h3><a class="anchor" id="file_menu"></a>
File Menu</h3>
<p>Load a pose through this menu first either from a file or directly from the RCSB. If you have PyMOL setup, it will send the pose to PyMOL after it loads into Rosetta. Loading a pose will open up the PDB Cleaning window by default to prep the PDB for import into Rosetta.</p>
<p>PDB Cleaning Functions: </p>
<ul>
<li>renaming waters </li>
<li>removing waters </li>
<li>removing hetatm </li>
<li>renaming DNA residues </li>
<li>renaming MD atoms </li>
<li>changing occupancies to 1.0 </li>
<li>renumbering pose from 1 + accounting for insertions </li>
<li>checking that all residues are able to be read by Rosetta</li>
</ul>
<ul>
<li>Current Imports:GUI Session, Param PathList, Loop File </li>
<li>Current Exports:GUI Session, Param PathList, Loop File, basic Resfile, basic Blueprint, SCWRL seq file, FASTA (pose + regions)</li>
</ul>
<h3><a class="anchor" id="main"></a>
Main Window</h3>
<p>The main window is used to run common protocols, some anlysis, setup decoy output, and finally set a region.</p>
<p>The default region is the whole protein. To specify loops for loop modeling, regions for minimization or residues for repacking, etc. add a region: </p>
<ul>
<li>Chain: can be added by just specifying the chain </li>
<li>N-terminus:Leave 'start' entry box blank, or enter any string, specify both chain and end residue number </li>
<li>C-terminus:Leave 'end' entry box blank, or enter any string, specify both chain and start residue number </li>
<li>Residue: Have start + end be the same residue.</li>
</ul>
<p>Sequence: </p>
<ul>
<li>A sequence of the pose or region can be found in the lower part of the screen. Placing your cursor at the left part of a residue will show its residue number, chain, and rosetta resnum. This can be colored in pymol and used to select residues in the Per-Residue Control Window described below.</li>
</ul>
<p>Output Options: </p>
<ul>
<li>The default output directory is the dir of the current pose. This and other output options for decoys can be set in the upper right and lower right areas of the main window. </li>
<li>Multiple Processors are used only for protocol runs of &gt;1 decoy, as well as some functions from the PDBList menu.</li>
</ul>
<p>Protocols: </p>
<ul>
<li>A few common minimization protocols can be ran through the main window in addition to the protocols menu (Repack, Minimize, Relax, SCWRL).</li>
</ul>
<h3><a class="anchor" id="options_menu"></a>
Options Menu</h3>
<ul>
<li>Option System manager: </li>
<li>: : Setup Rosetta command-line options using the options system manager </li>
<li>Configure Score Function: </li>
<li>: : Default Scorefunction is score12prime. The scorefunction set here is used by all protocols. See the protocols section for more information. Please use this window to change the scorefunction or the weights of any terms. </li>
<li>: : More weight sets can be given in this window by selecting options-&gt;show ALL scorefunctions </li>
<li>Set General Protocol Options: </li>
<li>: : A number of rounds can be set for all protocols. By default, if more than one round is set, the boltzmann criterion is applied to the pose after each round using the current scorefunction. This is especially useful for packing or design. Here, you can control the temperature, and the behavior of what happens each round and if the lowest energy pose is recovered at the end of the run.</li>
</ul>
<h3><a class="anchor" id="visualization_menu"></a>
Visualization Menu</h3>
<p>PyMOL Integration through the PyMOL Mover. A pose can be shown in PyMOL at any time. Regions can be colored, and the PyMOL Observer can be enabled that updates pymol upon structural and energetic changes to the current pose. </p>
<ul>
<li>Advanced PyMOL Visualization button </li>
<li>: : This is a simple window that can be used while using PyMOL. It is small so that you can incorporate it into your modeling workflow. </li>
<li>: : For energies, blue is low energy and red is high energy. Spectrum cannot be changed at this time, but will be added in the future. </li>
<li>: : Labels for energies can be useful, although slow. Use the PyMOL commands to remove them once sent.</li>
</ul>
<h3><a class="anchor" id="advanced_menu"></a>
Advanced Menu</h3>
<p>The Ligand/NCAA/PTM Manager </p>
<ul>
<li>Many non-cannonicals are off-by-default for Rosetta, and this window will allow you to enable them and also design residues into them. </li>
<li>Energy function optimization is also given here. Since the default electrostatic score term in the default scorefunction is statistical in nature, it will not work for non-cannoicals. At the very least, changing this to the coulumbic score term is a more physically realistic representation of the sytem. </li>
<li>The mm_std scorefunction ( Douglas Renfrew, Eun Jung Choi, Brian Kuhlman, 'Using Noncanonical Amino Acids in Computational Protein-Peptide Interface Design' (2011) PLoS One. ) is currently undergoing broad tests for proteins and other systems, but seems to work very will for NCAA design and rotamers. </li>
<li>The orbitals scorefunction, developed by Steven Combs et al (Meiler Lab) is also a great alternative, which is residue independant. Tests are ongoing for all of these scorefunction changes.</li>
</ul>
<p>Design Toolbox </p>
<ul>
<li>This is a window which will allow you to build a resfile for the loaded pose. You can do individual residues or a region of the pose. See the help menu or paper for information on what the values are in this window. </li>
<li>The left most box is for residue categories. This is for easy selection and design. Use the etc category for typical rosetta resfile options (NATRO,NATAA,ALLAA). Clicking one of the categories will populate the right box. Here you actually choose the residues types you want to open by double clicking them. Selecting these will limit the design to only those chosen. Residues you have chosen then go into the lower most box. Changing the residue number you have selected, will change what goes in this box so you can edit your current design. Double clicking a residue in the box will remove it. </li>
<li>Click the Write Resfile button to open a save as dialog.</li>
</ul>
<p>Enable constraints </p>
<ul>
<li>Will prompt for a constraint file. For more information, please see the rosettacommons manual.</li>
</ul>
<p>Insert data into B factor Column </p>
<ul>
<li>This is mainly an analysis utility, but it can be useful for data visualization, especially for publications. </li>
<li>Data should be in a text or csv file, deliminated by either a space, tab, column, comma. </li>
<li>Data should have at least 3 columns: pdb resnum, chain, data to insert, optional atom name </li>
<li>The UI for this utility function will allow you choose the colunns for each required piece as well as the deliminator. It will use your currently loaded pose and output a new pose with the data in the B-factor column. Load the pose into PyMOL or your favorite visualizer to see the data.</li>
</ul>
<p>Per Residue control and Analysis </p>
<ul>
<li>This is a great interactive modeling window. Here you can do many modeling tasks including mutating individual residues, add variants such as phosphorylations/acetylations, repack rotamers, chaing dihedrals, etc. You can also analze rotamer probabilies and individual residue energies.</li>
</ul>
<p>Analysis Movers: </p>
<ul>
<li>Interface Analyzer </li>
<li>Loop Analyzer </li>
<li>Packing Analyzer </li>
<li>VIP</li>
</ul>
<h3><a class="anchor" id="protocols_menu"></a>
Protocols Menu</h3>
<ul>
<li>A loaded pose is required for all protocols. Options not given in the main UI or their protocol-specific UI can be set using the options system manager. Please refer to RosettaCommons manual for all protocols, as well as the references listed there and within the GUI. </li>
<li>The default scorefunction is score12prime. Ligands/NCAA/PTM may require modifications to the scorefunction. Please use the advanced menu to make these modifications. Centroid-based protocols and other specific protocols require different scorefunctions. They are given below, and you will be prompted in the GUI if these are not set.</li>
</ul>
<p>Currently Implemented:</p>
<p>High/Low Resolution Docking </p>
<ul>
<li>Will prompt for docking partners using the same notation as the docking application. </li>
<li>ScoreFunction: Use interchain_cen scorefunction for low-resolution and docking for high-resolution.</li>
</ul>
<p>High/Low Resolution Loop Modeling (KIC/CCD) </p>
<ul>
<li>Will prompt for fragset(CCD)/ extend options. </li>
<li>Default cutpoint is around the middle of the loop. </li>
<li>ScoreFunction: Use centroid + score4L patch for low-resolution.</li>
</ul>
<p>Fixedbb Design (UI) </p>
<ul>
<li>Will prompt for a resfile.</li>
</ul>
<p>Floppytail (UI)</p>
<p>Grafting (UI) </p>
<ul>
<li>Graft a region from one protein into another. </li>
<li>Many options are given. All are described in the UI. Choosing a graft method will give a description of that method. </li>
<li>It is assumed that the loaded pose is the one you will graft into. You can pick the doner pose from the UI. They can both be the same PDB file. </li>
<li>If the region is not already deleted, it will delete it upon grafting.</li>
</ul>
<p>FastRelax</p>
<p>ClassicRelax</p>
<p>PackRotamers (Rosetta)</p>
<p>PackRotamers (SCWRL)</p>
<p>Minimizer </p>
<ul>
<li>The Rosetta default minimizer is dfpmin. Use the options system to set a different one setting '-run:min_tolerance' </li>
<li>The Relax application uses dfpmin_armijo_nonmonotone. Visit <a href="https://www.rosettacommons.org/manuals/archive/rosetta3.4_user_guide/d7/df1/minimization_overview.html">https://www.rosettacommons.org/manuals/archive/rosetta3.4_user_guide/d7/df1/minimization_overview.html</a> for more information. </li>
<li>The Rosetta default tolerance is 0.000001. Use the options system to set a different value by setting -run:min_tolerance.</li>
</ul>
<p>Server Links: </p>
<ul>
<li>ROSIE </li>
<li>Fragments </li>
<li>Backrub </li>
<li>Interface Alanine Scan </li>
<li>DNA Interface Scan </li>
<li>Scaffold Select</li>
</ul>
<h3><a class="anchor" id="pdblist_menu"></a>
PDBList Menu</h3>
<p>A PDBList is simply a list of full paths to PDB files, one on each line. It is used by the GUI to analyze decoys. Methods for making and using a PDBList reside here.</p>
<p>Main Functions: </p>
<ul>
<li>Lowest Energy - get lowest energy poses from a PDBList </li>
<li>Energy vs RMSD - Calculate Energy vs RMSD + output a file that can be graphed in any graphing application. </li>
<li>Output FASTA - Output FASTA with sequence for each pose or only of the given region </li>
<li>Design Statistics - Use FASTA to calculate design stats from a Rosetta design run. Outputs textfiles and an R plot. Needs reference pose loaded. Does not work with insertions and deletions during the design yet.</li>
</ul>
<h2><a class="anchor" id="tips"></a>
Tips</h2>
<ul>
<li>-Set default (command-line) Options through the options system manager window </li>
<li>-Set default Scorefunction through the scorefunction window - Used by all protocols. </li>
<li>-Save and Load GUI Sessions using the file menu </li>
<li>-Advanced Users: To add personal windows and functions to the GUI, see developer html in pyrosetta_toolkit/documentation directory. </li>
<li>Please visit <a href="http://bugs.rosettacommons.org">http://bugs.rosettacommons.org</a> for any Toolkit-specific bugs</li>
</ul>
<h3><a class="anchor" id="bachrc"></a>
Bashrc Setup</h3>
<p>This is so you do not need to source SetPyRosettaEnvironment.sh each time you want to use Toolkit. Takes a few minutes, but worth it. For more information on to what a bash shell is: <a href="http://superuser.com/questions/49289/what-is-bashrc-file">http://superuser.com/questions/49289/what-is-bashrc-file</a></p>
<p>To do this: </p>
<ul>
<li>1) open SetPyRosettaEnvironment.sh in a text editor (Can be found in PyRosetta directory). Copy it's text from where it says export PYROSETTA down. </li>
<li>2) open $HOME/.bashrc (linux) or $HOME/.bash_profile (mac) into vi or gedit through the command-line. --&gt; <code>vi $HOME/.bashrc</code> </li>
<li>3) paste what was in SetPyRosettaEnvironment, editing this line: <code>export PYROSETTA</code> to <code>export PYROSETTA='full/path/to/PyRosetta/directory'</code> </li>
<li>Also, Make sure there are no extra spaces next to the equals sign. </li>
<li>4) Your bashrc will run each time a new terminal window or tab is created. To run it now: source $HOME/.bashrc</li>
</ul>
<h3><a class="anchor" id="pymol"></a>
PyMOL Setup</h3>
<p>Setup so that the pymol server runs + listens each time PyMOL is opened.</p>
<ul>
<li>1) Add a line in the file $HOME/.pymolrc (may be a new file): <code>run full/path/to/PyMOLPyRosettaServer.py</code> (This file is found in the PyRosetta directory)</li>
</ul>
<h2><a class="anchor" id="changelog"></a>
Changes since last release</h2>
<p>This will be the first release </p>
</div></div><!-- contents -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>



<hr class="footer"/><address class="footer"><small>
Generated on Tue Apr 9 2013 13:23:32 for My Project by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.7.6.1
</small></address>

</body>
</html>
